File: reclaim/domain/states.py
"""Closed case-state set and the allowed transition table."""

from __future__ import annotations

from enum import Enum


class CaseState(str, Enum):
    NEW = "NEW"
    ENRICHING = "ENRICHING"
    DIAGNOSING = "DIAGNOSING"
    POLICY_EVAL = "POLICY_EVAL"
    ACTION_READY = "ACTION_READY"
    EXECUTING = "EXECUTING"
    AWAITING_CUSTOMER = "AWAITING_CUSTOMER"
    ATTEMPT_FAILED = "ATTEMPT_FAILED"
    AMBIGUOUS = "AMBIGUOUS"
    RECONCILING = "RECONCILING"
    HALTED = "HALTED"
    ESCALATED = "ESCALATED"
    VERIFIED_RECOVERED = "VERIFIED_RECOVERED"
    VERIFIED_FAILED = "VERIFIED_FAILED"
    EXPIRED_UNRESOLVED = "EXPIRED_UNRESOLVED"


CASE_STATES: tuple[CaseState, ...] = tuple(CaseState)

TERMINAL_STATES = frozenset(
    {
        CaseState.VERIFIED_RECOVERED,
        CaseState.VERIFIED_FAILED,
        CaseState.EXPIRED_UNRESOLVED,
    }
)

CLOCK_STOPPED_STATES = TERMINAL_STATES | {CaseState.HALTED}

# Explicit allowed edges. Absence is illegal.
ALLOWED_TRANSITIONS: frozenset[tuple[CaseState, CaseState]] = frozenset(
    {
        (CaseState.NEW, CaseState.ENRICHING),
        (CaseState.ENRICHING, CaseState.DIAGNOSING),
        (CaseState.ENRICHING, CaseState.ESCALATED),
        (CaseState.DIAGNOSING, CaseState.POLICY_EVAL),
        (CaseState.DIAGNOSING, CaseState.ESCALATED),
        (CaseState.POLICY_EVAL, CaseState.ACTION_READY),
        (CaseState.POLICY_EVAL, CaseState.ESCALATED),
        (CaseState.POLICY_EVAL, CaseState.VERIFIED_FAILED),
        (CaseState.ACTION_READY, CaseState.EXECUTING),
        (CaseState.ACTION_READY, CaseState.HALTED),
        (CaseState.EXECUTING, CaseState.AWAITING_CUSTOMER),
        (CaseState.EXECUTING, CaseState.ATTEMPT_FAILED),
        (CaseState.EXECUTING, CaseState.AMBIGUOUS),
        (CaseState.AWAITING_CUSTOMER, CaseState.VERIFIED_RECOVERED),
        (CaseState.AWAITING_CUSTOMER, CaseState.ATTEMPT_FAILED),
        (CaseState.AWAITING_CUSTOMER, CaseState.AMBIGUOUS),
        (CaseState.ATTEMPT_FAILED, CaseState.POLICY_EVAL),
        (CaseState.ATTEMPT_FAILED, CaseState.ESCALATED),
        (CaseState.AMBIGUOUS, CaseState.RECONCILING),
        (CaseState.AMBIGUOUS, CaseState.EXPIRED_UNRESOLVED),
        (CaseState.RECONCILING, CaseState.AWAITING_CUSTOMER),
        (CaseState.RECONCILING, CaseState.ATTEMPT_FAILED),
        (CaseState.RECONCILING, CaseState.AMBIGUOUS),
        (CaseState.RECONCILING, CaseState.EXPIRED_UNRESOLVED),
        (CaseState.HALTED, CaseState.ACTION_READY),
        (CaseState.HALTED, CaseState.EXPIRED_UNRESOLVED),
        (CaseState.ESCALATED, CaseState.EXECUTING),
        (CaseState.ESCALATED, CaseState.VERIFIED_FAILED),
        (CaseState.ESCALATED, CaseState.EXPIRED_UNRESOLVED),
    }
)

AUDIT_EVENT_TYPE = "state_transition"


def as_case_state(value: CaseState | str) -> CaseState:
    if isinstance(value, CaseState):
        return value
    return CaseState(value)


def is_allowed(expected: CaseState, new: CaseState) -> bool:
    return (expected, new) in ALLOWED_TRANSITIONS

File: web/src/lib/states.ts
/**
 * Case-state presentation.
 *
 * The vocabulary itself comes from the API (`/api/meta`), which reads it from
 * the domain. This module only says how a state should *look* and what it
 * means for an operator; it never invents a state or decides a transition.
 */

export type Semantic =
  | "success" | "attention" | "danger" | "progress" | "waiting" | "neutral";

interface StatePresentation {
  semantic: Semantic;
  /** What this state means for the operator, in one line. */
  meaning: string;
  /** True when a human is expected to act. */
  actionable?: boolean;
}

export const STATE_PRESENTATION: Record<string, StatePresentation> = {
  NEW:                { semantic: "neutral",  meaning: "Case opened, not yet enriched." },
  ENRICHING:          { semantic: "neutral",  meaning: "Gathering context for diagnosis." },
  DIAGNOSING:         { semantic: "neutral",  meaning: "Classifying the failure cause." },
  POLICY_EVAL:        { semantic: "neutral",  meaning: "Deciding what action is permitted." },
  ACTION_READY:       { semantic: "neutral",  meaning: "An action is approved and queued for dispatch." },
  EXECUTING:          { semantic: "progress", meaning: "Dispatching to the provider." },
  AWAITING_CUSTOMER:  { semantic: "waiting",  meaning: "Payment link is live; waiting on the customer." },
  ATTEMPT_FAILED:     { semantic: "neutral",  meaning: "This attempt failed; the case may try another action." },
  // Unknown money state is an operational emergency, not a quiet grey row.
  AMBIGUOUS:          { semantic: "attention", actionable: true,
                        meaning: "The provider outcome is unknown. Money may or may not have moved." },
  RECONCILING:        { semantic: "progress", meaning: "Querying the provider to resolve the unknown outcome." },
  HALTED:             { semantic: "danger",   actionable: true,
                        meaning: "Dispatch stopped by the circuit breaker." },
  ESCALATED:          { semantic: "attention", actionable: true,
                        meaning: "Handed to a human for a decision." },
  VERIFIED_RECOVERED: { semantic: "success",  meaning: "Payment independently verified. Revenue recognised." },
  // A legitimate resolved outcome, not a system error — so not red.
  VERIFIED_FAILED:    { semantic: "neutral",  meaning: "Confirmed not recovered. Case closed." },
  EXPIRED_UNRESOLVED: { semantic: "attention", actionable: true,
                        meaning: "Time ran out before the outcome could be resolved." },
};

export function presentationFor(state: string): StatePresentation {
  return (
    STATE_PRESENTATION[state] ?? {
      semantic: "neutral",
      meaning: "Unrecognised state.",
    }
  );
}

/** Human-facing label. The literal state name is always shown alongside. */
export function humanizeState(state: string): string {
  return state
    .toLowerCase()
    .split("_")
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(" ");
}
